optuna/pruners/hyperband.py

```diff
@@ -150,21 +150,31 @@ def prune(self, study: "optuna.study.Study", trial: FrozenTrial) -> bool:
         bracket_study = self._create_bracket_study(study, i)
         return self._pruners[i].prune(bracket_study, trial)
 
-    # TODO(crcrpar): Improve resource computation/allocation algorithm.
-    def _calc_bracket_resource_budget(self, pruner_index: int, n_brackets: int) -> int:
-        n = self._reduction_factor ** (n_brackets - 1)
-        return n + (n / 2) * (n_brackets - 1 - pruner_index)
+    def _calculate_trial_allocation_budget(self, pruner_index: int) -> int:
+        """Compute the trial allocated budget for a bracket of ``pruner_index``.
+
+        In the `original paper <http://www.jmlr.org/papers/volume18/16-558/16-558.pdf>`, the
+        number of trials per one bracket is referred as ``n`` in Algorithm 1. Since we do not know
+        the total number of trials in the leaning scheme of Optuna, we calculate the ratio of the
+        number of trials here instead.
+        """
+
+        s = self._n_brackets - 1 - pruner_index
+        return math.ceil(self._n_brackets * (self._reduction_factor ** s) / (s + 1))
 
     def _get_bracket_id(self, study: "optuna.study.Study", trial: FrozenTrial) -> int:
-        """Computes the index of bracket for a trial of ``trial_number``.
+        """Compute the index of bracket for a trial of ``trial_number``.
 
         The index of a bracket is noted as :math:`s` in
         `Hyperband paper <http://www.jmlr.org/papers/volume18/16-558/16-558.pdf>`_.
         """
 
-        n = hash("{}_{}".format(study.study_name, trial.number)) % self._resource_budget
+        n = (
+            hash("{}_{}".format(study.study_name, trial.number))
```

Comment:
[Just FYI and it's not relevant to this PR itself]
We might be able to simplify this logic as, for instance, `hash(trial.number)`.

Reply:
Hello from 2021!
I think there could be an error about using hash.
To be specific, I conducted the parallel Optuna experiments using the Hyperband Pruner.
Then, the hash gives different value for the same input when executed on different terminal (different python script). For example, a trial _n_ could be assigned to the bracket 1 at process 1 and the bracket 2 at process 2 at the same time, which ruins the benefits of Hyperband over the SuccesiveHalving.
As you can see [here](https://docs.python.org/3.3/using/cmdline.html), it was due to the update that hash randomization is turned on after python version 3.3.
Of course, the randomization could be turned off by setting the _PYTHONHASHSEED=0_ when executing the python script, but I did not see any warning or suggestion about it at the documentation.
Since there are some papers claiming that utilizing Optuna parallelization with the Hyperband pruner improves the performance, I would like to ask if my assertion was legit.
I include the screenshots that the trials belonging to the same bracket at each parallel process were computed differently.
 ![Process 1 - trials for bracket 1](https://user-images.githubusercontent.com/55653250/140302703-a7ea8e8a-0f9d-4be1-ae1c-246686535c17.png)

![Process 2 - trials for bracket 1](https://user-images.githubusercontent.com/55653250/140302720-94e346e1-0984-4278-a278-005534efe9cc.png)